lib/Conversion/ExportVerilog/HWLowerInstanceChoices.cpp

```diff
@@ -0,0 +1,126 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the "prepare" pass that walks the IR before the emitter
+// gets involved.  This allows us to do some transformations that would be
+// awkward to implement inline in the emitter.
+//
+// NOTE: This file covers the preparation phase of `ExportVerilog` which mainly
+// legalizes the IR and makes adjustments necessary for emission. This is the
+// place to mutate the IR if emission needs it. The IR cannot be modified during
+// emission itself, which happens in parallel.
+//
+//===----------------------------------------------------------------------===//
+
+#include "../PassDetail.h"
+#include "ExportVerilogInternals.h"
+#include "circt/Conversion/ExportVerilog.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Support/LLVM.h"
+#include "circt/Support/Namespace.h"
+#include "circt/Support/SymCache.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/IR/Threading.h"
+#include "llvm/Support/Debug.h"
+
+#include <mutex>
+
+#define DEBUG_TYPE "lower-hw-instance-choices"
+
+using namespace mlir;
+using namespace circt;
+using namespace hw;
+using namespace sv;
+using namespace ExportVerilog;
+
+namespace {
+
+struct HWLowerInstanceChoicesPass
+    : public HWLowerInstanceChoicesBase<HWLowerInstanceChoicesPass> {
+  void runOnOperation() override;
+};
+
+} // end anonymous namespace
+
+LogicalResult
+ExportVerilog::lowerHWInstanceChoices(mlir::ModuleOp module,
+                                      const LoweringOptions &options) {
+  // Collect all instance choices & symbols.
+  SmallVector<InstanceChoiceOp> instances;
+  SymbolCache symCache;
+  {
+    for (Operation &op : *module.getBody()) {
+      if (auto module = dyn_cast<HWModuleOp>(&op)) {
+        module.walk([&](InstanceChoiceOp inst) { instances.push_back(inst); });
+        symCache.addSymbol(module);
+        continue;
+      }
+      if (auto sym = dyn_cast<SymbolOpInterface>(&op)) {
+        symCache.addSymbol(sym);
+        continue;
+      }
+    }
+  }
+
+  // Build a namespace to generate unique macro names.
+  Namespace ns;
+  ns.add(symCache);
+
+  auto declBuilder = OpBuilder::atBlockBegin(module.getBody());
+  for (InstanceChoiceOp inst : instances) {
+    auto parent = inst->getParentOfType<HWModuleOp>();
+
+    auto defaultModuleOp = cast<HWModuleLike>(
+        symCache.getDefinition(inst.getDefaultModuleNameAttr()));
+
+    // Generate a macro name to describe the target of this instance.
+    std::string name;
+    {
+      llvm::raw_string_ostream os(name);
+      os << "__circt_choice_" << parent.getName() << "_"
+         << inst.getInstanceName();
+    }
```

Comment:
As stated above, nice scope to hide `os` from unnecessarily polluting the outer namespace.

Reply:
That said, please have the PR favor `SmallString` as this is more aligned with the LLVM Developer Policy and would avoid a heap allocation as long as the string is below the pre-allocated size.